### PR TITLE
fetch: Add row count per file to resources

### PR DESCRIPTION
**Pull request by Jeremyyang920**
This commit adds a row count field to ever storage resource. The reason for this is mainly for progress tracking but more specfically for the IMPORT use case where we are not able to know the number of rows imported. We now have this information since each resource has its own file count.

Release note: None

### CODE REVIEW

**Comment by ryanluu12345** [comment]
Could you please add results of your manual testing with each fetch mode? I mainly want to see it hasn't regressed. Or even better, run the E2E tests on this branch?

File: fetch/csv_pipe.go [hunk 1]
```diff
@@ -79,8 +85,14 @@ func (p *csvPipe) Pipe(tn dbtable.Name) error {
 	}
 }
 
+// Flush flushes the current csv files when either
```
Comment (ryanluu12345) [inline fetch/csv_pipe.go:88]: Super nit: `// flush` since the method is lower cased

File: fetch/csv_pipe.go [hunk 2]
```diff
@@ -79,8 +85,14 @@ func (p *csvPipe) Pipe(tn dbtable.Name) error {
...
+// we reached the end of the file or our file limits.
+// It also sends the current number of rows processed
+// to a channel to be processed by the data storage
+// backend.
 func (p *csvPipe) flush() error {
 	if p.csvWriter != nil {
+		p.numRowsCh <- p.currRows
```
[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: If the flush fails then we exit the program anyways so the ordering here doesn't really matter.

File: fetch/csv_pipe.go [hunk 3]
```diff
@@ -22,25 +22,31 @@ type csvPipe struct {
 	currSize  int
 	currRows  int
 	numRows   int
-	newWriter func() io.WriteCloser
+	numRowsCh chan int
+	newWriter func(numRowsCh chan int) io.WriteCloser
```
[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:26]: Because of the way newWriter is defined as a closure on the csvPipe object and where file resources get created, a channel has to be used because its the only form of way of updating data to this go routine. Its also the most idomatic way of passing data between go routines.

File: fetch/csv_pipe.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZhouXing19) [reply fetch/csv_pipe.go:95]: I also found a mismatch here. 

In `csvPipe.flush()`, we first `p.numRowsCh <- p.currRows` and then `p.csvWriter.Flush()`, while in the parallel goroutine, we have `gcpStore.CreateFromReader()` first call `io.Copy(wc, r);` and then `rows := <-numRows`. IIUC `p.numRowsCh` is a blocking channel, so it actually seems a deadlock to me. Thus I wonder if we should call `rows := <-numRows` before `io.Copy(wc, r);` in `gcpStore.CreateFromReader()`.

[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: p.numRowsCh is not blocking as its buffered.

[the remark this answers is not in the document]
Comment (ZhouXing19) [reply fetch/csv_pipe.go:95]: Ok makes sense that it is not a deadlock, but if error happens `CreateFromReader()` returns with err prior to `rows := <-numRows`, the _next_ iteration of `p.flush()` in the for-loop in `Pipe()` will still be blocked, as the `numRows` is still full, making the main program hang at `err := pipe.Pipe(table.Name)` in `func exportTable()`.

[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: If im not mistaken we exit the file creation process at that point, the error from p.flush() the error channels passes it along and we exit this process completely.

[the remark this answers is not in the document]
Comment (ZhouXing19) [reply fetch/csv_pipe.go:95]: > the error from p.flush() the error channels passes it along 

The error is pushed to `writerErrCh` in `exportTable()`, but `writerErrCh` is read only after `pipe.Pipe()` returns. The case I had in mind is that `pipe.Pipe()` hangs, and `writerErrCh` will never be read.

[the remark this answers is not in the document]
Comment (ZhouXing19) [reply fetch/csv_pipe.go:95]: The story line I had in mind is:
1. In the 1st iteration of the for-loop in `csvPipe.Pipe()`, we call `p.flush()`, which do `p.numRowsCh <- p.currRows; p.csvWriter.Flush()`, and return immediately. i.e. we will enter the 2nd iteration immediately. 
2. In parallel, the `gcpStore.CreateFromReader()` returned with error at `io.Copy(wc, r)`, without emptying  `numRowsCh`. The error is written to `writerErrCh` in `exportTable()`
3. In the 2nd iteration, `p.flush()` now hangs, as `p.numRowsCh` is still full. It means `pipe.Pipe()` hangs.
4. At this moment, back to the main program -- `exportTable()` --, it hangs at `err := pipe.Pipe(table.Name)`, while `case werr := <-writerErrCh:` actually is unreachable.

[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: Sorry yes, you are correct. So its actually just an ordering thing of reading off the channel. The issue I was going to describe was a thing I saw with S3 and why I had the read after, but because I send the data first in csv_pipe before flush, the ordering of the read inside the CreateFromReader() call won't matter and I can do that before possible errors.

[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: But looking through this thread now, we need to add a test case around more errors since none of our existing tests would have caught a this.

[the remark this answers is not in the document]
Comment (ryanluu12345) [reply fetch/csv_pipe.go:95]: Yes I agree, we should add some datadriven tests for this case. Or if not appropriate, can add tests that simulate these cases and verify deadlocks/timeouts don't occur.

[the remark this answers is not in the document]
Comment (ryanluu12345) [reply fetch/csv_pipe.go:95]: Based on what Jane is describing for the ordering of reading off the channel, does that mean we just need to read the error of the channel before the numRows? Would that end up working?

**Review by ryanluu12345 — CHANGES_REQUESTED** [review]
Pending the ordering changes mentioned in the thread to avoid blocking behavior.

File: fetch/csv_pipe.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZhouXing19) [reply fetch/csv_pipe.go:95]: Thinking about it more, I think there's a bug that is here originally and a better exit rule for Pipe() is needed -- the writer creation part should be able to communicate with it if error happens. I will need some time to think and test more to confirm this tho...

[the remark this answers is not in the document]
Comment (Jeremyyang920) [reply fetch/csv_pipe.go:95]: TLDR: Its note even just the ordering of things really that is the problem. Io.Pipe operations with copy are blockng operations that can also prevent channel processing. So there is a delicate balance here of ensuring things don't block when its not supposed to.

**Comment by Jeremyyang920** [comment]
Closing this pr as Jane will take it across the line with the bug fixes since the row count isint urgent.

**Comment by ZhouXing19** [comment]
Note here as well: implementation from this PR is merged with #125 but that PR doesn't include tests for the original row count add-on for each blob storage, it only includes test for the deadlock issue. The test is tracked in this jira ticket: https://cockroachlabs.atlassian.net/browse/CC-27091
